### DISCUSSION TITLE
[sqlalchemy related] Session.refresh vs self.execute to refresh data

### DISCUSSION THREAD

**Discussion by cemrehancavdar** [sqlalchemy/sqlalchemy#11554] [Usage Questions] [answered]
Hi,

We're using Advanced Alchemy to handle database actions, which provides an API for specifying loading techniques at the repository level. The repository adds these techniques to the statement before executing a select query.

However, we've encountered an issue with the "loading techniques." The refresh function does not consider relationship loading techniques. Consequently, every time we create or update an entry, we need to refetch the data.

Advanced Alchemy uses AsyncSession.refresh to refetch data after performing update or create actions. I am considering to suggest to Advanced Alchemy "replacing all refresh calls with .get calls", but I want to ensure this is the correct approach.

From what I understand, the refresh function is basiscally select query. Is it safe to say "use an execute statement to select data with options instead"? Do i need to know something more? Could you please share your insights on this matter?

https://github.com/litestar-org/advanced-alchemy/blob/d3f8cfd8b95fd4b4c98348d6387d9ea7f6c708ad/advanced_alchemy/repository/_async.py#L1157

Thank you!

**Comment by zzzeek** [comment, marked as answer]
1. is this a SQLAlchemy question or an alembic question?
2. I'm not familiar with what advanced alchemy is trying to achieve but I would not recommend a global, across the board get or refresh for a SQLAlchemy application, proactively performing a per-object SELECT after every operation is very wasteful.    I would never use that approach, if I'm using refresh() or a get() for refreshing, it's going to be explicit for a particular object that my application definitely needs to be refreshed in this way (and I would prefer to not have that needed since I always want to avoid one-off SELECTs for a single row if I can
3.   In regular SQLAlchemy we suggest people turn off object expiration if they want to avoid objects that dont work without a refresh, not sure what this library is doing that changes that
4. session.refresh() will restore relationship based attributes that were eagerly loaded as well, and can also eagerly load relationships that were lazy loaded if they are named in the attributes collection.   so there should not be much difference between using session.get().

**Reply by cemrehancavdar** [reply]
Didn't notice i was on alembic's repo, sorry about that. Your answer is enough for me, thanks.

**Reply by CaselIT** [reply]
(moved to sqlalchemy repo)
